### ISSUE TITLE
Shared directories are failing

### ISSUE THREAD

**Issue by trsteel88** [EasyCorp/easy-deploy-bundle#22]
Deployment keeps failing with shared directories. Config below.

```
Initializing configuration
Processing the configuration options of the deployer class
binDir                     : app
cacheDir                   : app/cache
composerInstallFlags       : --no-dev --prefer-dist --no-interaction --quiet
composerOptimizeFlags      : --optimize
configDir                  : app/config
consoleBinaryPath          :
controllersToRemove        : []
deployDir                  : /home/deploydemo/application
dumpAsseticAssets          : true
installWebAssets           : true
keepReleases               : 2
localProjectDir            : /usr/local/var/www/htdocs/deploydemo
logDir                     : app/logs
permissionGroup            :
permissionMethod           : chmod
permissionMode             : 0777
permissionUser             :
remoteComposerBinaryPath   : /usr/local/bin/composer
remotePhpBinaryPath        : php
repositoryBranch           : master
repositoryUrl              : xxxxxxxx
resetOpCacheFor            :
servers                    : deploydemo@domain.com
sharedDirs                 : ["app/logs/"]
sharedFiles                : ["app/config/parameters.yml"]
srcDir                     : src
symfonyEnvironment         : staging
templatesDir               : app/Resources/views
updateRemoteComposerBinary : false
useSshAgentForwarding      : false
warmupCache                : true
webDir                     : web
writableDirs               : ["app/cache/","app/logs/"]
```

```
Preparing app
Creating cache directory
[deploydemo@domain.com] Executing command: (export SYMFONY_ENV=staging; cd /home/deploydemo/application/releases/20170615103114 && if [ -d /home/deploydemo/application/releases/20170615103114/app/cache ]; then rm -rf /home/deploydemo/application/releases/20170615103114/app/cache; fi; mkdir -p /home/deploydemo/application/releases/20170615103114/app/cache)
Creating log directory
[deploydemo@domain.com] Executing command: (export SYMFONY_ENV=staging; cd /home/deploydemo/application/releases/20170615103114 && if [ -d /home/deploydemo/application/releases/20170615103114/app/logs ] ; then rm -rf /home/deploydemo/application/releases/20170615103114/app/logs; fi; mkdir -p /home/deploydemo/application/releases/20170615103114/app/logs)
Creating symlinks for shared directories
[deploydemo@domain.com] Executing command: (export SYMFONY_ENV=staging; cd /home/deploydemo/application/releases/20170615103114 && mkdir -p /home/deploydemo/application/shared/app/logs/)
[deploydemo@domain.com] Executing command: (export SYMFONY_ENV=staging; cd /home/deploydemo/application/releases/20170615103114 && if [ -d /home/deploydemo/application/releases/20170615103114/app/logs/ ] ; then rm -rf /home/deploydemo/application/releases/20170615103114/app/logs/; fi)
[deploydemo@domain.com] Executing command: (export SYMFONY_ENV=staging; cd /home/deploydemo/application/releases/20170615103114 && ln -nfs /home/deploydemo/application/shared/app/logs/ /home/deploydemo/application/releases/20170615103114/app/logs/)
| err :: ln: target `/home/deploydemo/application/releases/20170615103114/app/logs/' is not a directory: No such file or directory
[ERROR] Cancelling the deployment and reverting the changes
A log file with all the error details has been generated in /usr/local/var/www/htdocs/deploydemo/app/logs/deploy_prod.log
Executing beforeCancelingDeploy hook
```

**Comment by joelclouddistrict** [comment]
I am having the same issue:

```
===> Initializing configuration
---> Processing the configuration options of the deployer class
binDir                     : app
cacheDir                   : app/cache
composerInstallFlags       : --no-dev --prefer-dist --no-interaction --quiet
composerOptimizeFlags      : --optimize
configDir                  : app/config
consoleBinaryPath          : 
controllersToRemove        : ["web/app_dev.php"]
deployDir                  : /home/deploy/app
dumpAsseticAssets          : true
installWebAssets           : true
keepReleases               : 5
localProjectDir            : /application
logDir                     : app/logs
permissionGroup            : 
permissionMethod           : chmod
permissionMode             : 0777
permissionUser             : 
remoteComposerBinaryPath   : /usr/local/bin/composer
remotePhpBinaryPath        : php
repositoryBranch           : master
repositoryUrl              : xxxxx
resetOpCacheFor            : 
servers                    : xxxxx
sharedDirs                 : ["app/logs/"]
sharedFiles                : ["app/config/parameters.yml"]
srcDir                     : src
symfonyEnvironment         : prod
templatesDir               : app/Resources/views
updateRemoteComposerBinary : false
useSshAgentForwarding      : true
warmupCache                : true
webDir                     : web
writableDirs               : ["app/cache/","app/logs/"]
```

```
Preparing app
Creating cache directory
[xxxxx(remote)] Executing command: (export SYMFONY_ENV=prod; cd /home/deploy/app/releases/20171023104827 && if [ -d /home/deploy/app/releases/20171023104827/app/cache ]; then rm -rf /home/deploy/app/releases/20171023104827/app/cache; fi; mkdir -p /home/deploy/app/releases/20171023104827/app/cache)
Creating log directory
[xxxxx(remote)] Executing command: (export SYMFONY_ENV=prod; cd /home/deploy/app/releases/20171023104827 && if [ -d /home/deploy/app/releases/20171023104827/app/logs ] ; then rm -rf /home/deploy/app/releases/20171023104827/app/logs; fi; mkdir -p /home/deploy/app/releases/20171023104827/app/logs)
Creating symlinks for shared directories
[xxxxx(remote)] Executing command: (export SYMFONY_ENV=prod; cd /home/deploy/app/releases/20171023104827 && mkdir -p /home/deploy/app/shared/app/logs/)
[xxxxx(remote)] Executing command: (export SYMFONY_ENV=prod; cd /home/deploy/app/releases/20171023104827 && if [ -d /home/deploy/app/releases/20171023104827/app/logs/ ] ; then rm -rf /home/deploy/app/releases/20171023104827/app/logs/; fi)
[xxxxx(remote)] Executing command: (export SYMFONY_ENV=prod; cd /home/deploy/app/releases/20171023104827 && ln -nfs /home/deploy/app/shared/app/logs/ /home/deploy/app/releases/20171023104827/app/logs/)
| err :: ln: el objetivo «/home/deploy/app/releases/20171023104827/app/logs/» no es un directorio: No existe el archivo o el directorio
[ERROR] Cancelling the deployment and reverting the changes
A log file with all the error details has been generated in /application/app/logs/deploy_dev.log
```

The remote host is an Ubuntu 14.04.5

**Comment by patrick-mota** [comment]
I have the same problem.
Remote host: CentOS 7

I did "watch ls" to see the exact moment when the release is created then I canceled the deployment to check if the folder was on app/ and in fact it was....

I will try now to reproduce manually every single command from easydeploy.

**Comment by patrick-mota** [comment]
Like I said before I tried manually to reproduce every single command.
I was wrong in one of my assumption. The logs folder was there but not in releases/....

So, I found the solution.

Just create a folder logs in your git repository (app/logs), because in the release folder you don't get the code from your host but from git (isn't uploaded but taken from git repository).
That's why you need to have a logs folder.

The step which we got an error is this one:
```
-sh-4.1$ export SYMFONY_ENV=prod; 
cd /var/www/app/myproject/releases/20171204152139 && 
ln -nfs /var/www/app/myproject/releases/20171204152139/shared/app/logs/ 
/var/www/app/myproject/releases/20171204152139/app/logs/

ln: target '/var/www/app/myproject/releases/20171204152139/app/logs/' 
is not a directory: No such file or directory
```

The fix : have a folder in app/logs
I created the folder manually just to try it out but the "clean" solution is to put in your git repository the logs folder.

```
-sh-4.1$ cd /var/www/app/myproject/releases/20171204152139/
-sh-4.1$ mkdir logs
-sh-4.1$ export SYMFONY_ENV=prod; 
cd /var/www/app/myproject/releases/20171204152139 && 
ln -nfs /var/www/app/myproject/releases/20171204152139/shared/app/logs/ 
/var/www/app/myproject/releases/20171204152139/app/logs/
-sh-4.1$

```
As you can see no error.

Feel free to reply if you don't have the problem anymore so we can put this as solved.

Thank you

**Comment by patrick-mota** [comment]
@moderator feel free to delete my 2 old posts, I'm not sure if I should keep them here for information even if it's wrong...

Well, I'm still having the issue when executing the script.
The problem is in \EasyCorp\Bundle\EasyDeployBundle\Deployer\DefaultDeployer
doCreateShareDirs function
Line 285 you do:
            $this->runRemote(sprintf('if [ -d {{ project_dir }}/%s ] ; then rm -rf {{ project_dir }}/%s; fi', $sharedDir, $sharedDir));

The problem is you remove the app/logs/ folder so we can't do the ln -nfs right after.
I just added right after this line (for testing purposes only) 
            $this->runRemote('mkdir -p app/logs');
and it seems to fix the issue since the deploy continue until another error (php memory limit because of composer so completly not related).

**Comment by javiereguiluz** [comment]
Fixed by #37.
